extensions/shared/register-jetpack-block.js

```diff
@@ -32,12 +47,13 @@ export default function registerJetpackBlock( name, settings, childBlocks = [] )
 		return false;
 	}
 
-	const result = registerBlockType(
-		`jetpack/${ name }`,
-		betaExtensions.includes( name )
-			? { ...settings, title: `${ settings.title } (beta)` }
-			: settings
-	);
+	const result = registerBlockType( `jetpack/${ name }`, {
+		...settings,
+		title: betaExtensions.includes( name ) ? `${ settings.title } (beta)` : settings.title,
+		edit: requiredPlan
+			? wrapPaidBlock( settings.edit, { feature: blockFeatureMap[ name ], requiredPlan } )
```

Comment:
I see two ways of going about this. This way, were we have a nexus all the blocks pass through and we can chose to wrap them or not based on given conditions. This is appealing because we write code in one place (and make changes to the PHP-side unavailability), and it works for all the blocks.

The other way I was considering was to wrap blocks where they are defined. We'd describe the conditions for rendering the nudge or the plain block in individual block definitions and the HOC would need to be a bit smarter. The section of code here would remain mostly unchanged.

```js
const blockSettings = {
  // …snip
  edit: requirePremiumPlan( edit ),
  save,
}
```

I'm wondering what the tradeoffs are for one or another 🤔 

Given that these only touch `edit`, I don't think there are concerns related to invalidation or if blocks change plans. Did you consider this alternative or have thoughts about the tradeoffs?

Reply:
Might consider also encapsulating the plan display / check in a hook.

Reply:
> I see two ways of going about this. This way, were we have a nexus all the blocks pass through and we can chose to wrap them or not based on given conditions. This is appealing because we write code in one place (and make changes to the PHP-side unavailability), and it works for all the blocks.
> 
> The other way I was considering was to wrap blocks where they are defined. We'd describe the conditions for rendering the nudge or the plain block in individual block definitions and the HOC would need to be a bit smarter. The section of code here would remain mostly unchanged.
> 
> ```js
> const blockSettings = {
>   // …snip
>   edit: requirePremiumPlan( edit ),
>   save,
> }
> ```
> 
> I'm wondering what the tradeoffs are for one or another 
> 
> Given that these only touch `edit`, I don't think there are concerns related to invalidation or if blocks change plans. Did you consider this alternative or have thoughts about the tradeoffs?

I considered adding the nudge wrap to blocks individually, yeah. I decided to use the "nexus" approach since it means that it'll work for any block out of the box, provided that we'll provide the correct availability information through the endpoint. (It felt like a duplicated effort to add that information on the backend, _and_ add invidual block handling on the frontend. I'm assuming here that we _are_ going to add that information to the availability endpoint because it just seems to be the right thing to do.)

Furthermore, requiring block authors to manually add the wrapper to their blocks means there's a risk of copy-past gone wrong, or getting outdated.

That said, I hope this doesn't come across as defensive of my decision -- I don't really feel _that_ strongly about it, just trying to explain my decision. There is a point in having block registration _just_ register blocks, and move handling of 'special' cases to individual blocks :man_shrugging: 

> Might consider also encapsulating the plan display / check in a hook.

Yeah, that's what @scruffian originally had in https://github.com/Automattic/wp-calypso/pull/33723. I'd say that similar arguments apply as for @sirreal's suggestion. Compared to Jon's suggestion, filters might be a bin more semantic, but obvisously also a bit harder to locate.

Reply:
Thanks for the detail, I'm happy to proceed as implemented 👍